gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java

```diff
@@ -272,24 +326,102 @@ protected BoundField(String name, boolean serialized, boolean deserialized) {
+    private final ObjectConstructor<T> constructor;
+
+    FieldReflectionAdapter(ObjectConstructor<T> constructor, Map<String, BoundField> boundFields) {
+      super(boundFields);
+      this.constructor = constructor;
+    }
+
+    @Override
+    T createIntermediary() {
+      return constructor.construct();
+    }
+
+    @Override
+    void readField(T intermediary, JsonReader in, BoundField field)
+        throws IllegalAccessException, IOException {
+      field.read(in, intermediary);
+    }
+
+    @Override
+    T finalize(T intermediary) {
+      return intermediary;
+    }
+  }
+
+  private static final class RecordAdapter<T> extends Adapter<T, Object[]> {
+    // The actual record constructor.
+    private final Constructor<? super T> constructor;
+    // Array of arguments to the constructor, initialized with default values for primitives
+    private final Object[] constructorArgsDefaults;
+    // Map from field names to index into the constructors arguments.
+    private final Map<String, Integer> fieldIndices = new HashMap<>();
+
+    RecordAdapter(Class<? super T> raw, Map<String, BoundField> boundFields) {
+      super(boundFields);
+      this.constructor = ReflectionHelper.getCanonicalRecordConstructor(raw);
+      String[] recordFields = ReflectionHelper.recordFields(raw);
+      for (int i = 0; i < recordFields.length; i++) {
+        fieldIndices.put(recordFields[i], i);
+      }
+      Class<?>[] parameterTypes = constructor.getParameterTypes();
+      constructorArgsDefaults = new Object[parameterTypes.length];
+      for (int i = 0; i < parameterTypes.length; i++) {
+        if (parameterTypes[i].isPrimitive()) {
+          constructorArgsDefaults[i] = Array.get(Array.newInstance(parameterTypes[i], 1), 0);
+        }
       }
+    }
 
-      out.beginObject();
+    @Override
+    Object[] createIntermediary() {
+      Object[] constructorArgs = new Object[constructorArgsDefaults.length];
+      System.arraycopy(constructorArgsDefaults, 0, constructorArgs, 0, constructorArgs.length);
+      return constructorArgs;
+    }
+
+    @Override
+    void readField(Object[] intermediary, JsonReader in, BoundField field) throws IOException {
+      Integer fieldIndex = fieldIndices.get(field.name);
+      if (fieldIndex == null) {
+        throw new IllegalStateException(
+            "Could not find the index in the constructor "
+                + constructor
+                + " for field with name "
+                + field.name
+                + ", unable to determine which argument in the constructor the field corresponds"
+                + " to. This is unexpected behaviour, as we expect the RecordComponents to have the"
+                + " same names as the fields in the Java class, and that the order of the"
+                + " RecordComponents is the same as the order of the canonical arguments.");
+      }
+      intermediary[fieldIndex] = field.read(in);
```

Comment:
Might have to handle the case where the JSON contains an explicit `null` for a primitive record component. Maybe similar to `ReflectiveTypeAdapterFactory`'s current behavior, the `null` should be ignored in that case:
https://github.com/google/gson/blob/2591ede59bf7e4cef496a01e82282c88e256cec8/gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java#L162

Reply:
Fixed